### ISSUE TITLE
[Accessibility] Change how screen names are presented

### ISSUE THREAD

**Issue by pitermach** [DagAgren/toot#382]
Currently, Toot! announces a user's full screen name (@username@instance.url) after their full name on every toot. Additionally, there's some character substitution going on in the instance URL where any periods are replaced with " dot ".

This aproach makes reading toots inefficient, because you have to wait for the speech to read the full name every time before it gets to the actual toot text. With a braille display, this is even worse. A braille display can only display a small number of characters at a time (most common are 14, 20 or 40-character long displays) and you have to scroll the line to continue reading. With the way Toot! presents my username "@pitermach@mastodon dot social), not counting the parentheses it takes up 30 characters, meaning when I'm using a 14-cell display I literarly have to scroll through 2 lines of screen name just to get to the toot text.

So, I'd like to propose 2 changes.

- Don't put the screen name after the full name. Both the official mastodon.social web interface and the [Pinafore](https://pinafore.social/) web app put the screen name at the end, after the timestamp. Additionally, if the person is in the same instance as you then the domain part isn't included.
- Don't substitute the periods. Speech is usually smart enough to read them as a dot if it's part of a URL, and this way it takes up only 1 character in braille as well as if using the rotor to move by character. It will also localise better this way, as if I switch to a non-english voice it'll use the correct word for dot.

**Comment by DagAgren** [comment]
Unfortunately, in testing, speech is not at all smart enough to do that for most instance names, because they don't look like domain names, and it will read it as a full stop and go "mastodon (long pause) social", which is why the substitution is being done.

Also, I am torn on how to deal with names. Because the screen name is how the person has decided to present themselves, so it seems rude to leave it out, and putting it at the end seems very awkward. But people also change these often, so you can't leave out the username.

I didn't know you can use braille readers with iPhones, though, how does that work?

**Comment by pitermach** [comment]
I understand the centiment regarding the screen names. The problem with leaving it at the beginning though is that speech is linear, so it always reads through everything left to right. Visually, you can skim past it easily because it's written with a different font so you can just move your eyes past it and concentrate on the text, but someone using speech or braille can't do that because there isn't a "fast forward" button and constantly hearing something like "John smith, at JohnSmith42 at cool dot domain" before every toot gets old pretty quickly. And while I agree that someone's screen name is important, it's probably not as important as the actual toot especially if proceeded by a user's display name.

Looking through closed issues, I see that @MarcoZehe in #123 proposed a similar format to what I'm suggesting, with the full name followed by the toot text. It's also what other apps have seddled on, not just Mastodon clients.

Regarding dot substitution and Braille, yep, you can use a braille display to not only read what's on the screen, but also take full control of the device, with displays having buttons that do equivllent commands to various screen gestures as well as a keyboard to type text. Here's a [video of such a display in action](https://www.youtube.com/watch?v=19zVFYOA30g).

On a technical note, anything you put in the accessibility name gets rendered in braille on a display. As far as I know there is sadly no way to provide a different text for speech and braille independently of each other. So, any substitutions you do for speech will also effect braille output. So while a period normally takes up a single character, doing the substitution literarly causes the word "dot" to be spelled out in braille instead, taking up more space.

VoiceOver also has punctuation settings that control what characters are spoken and how they're read. This level is set to "some" by default, but users can change it as well as alter how individual symbols are spoken. I personally left it it on some, but I know people who have the level on none and prefer things like dots in URL's to be suppressed all the time.

Finally, if I want to review text in detail, VoiceOver has the rotor (doing a 2-finger rotate gesture, like turning a dial lets you select the type of movement, by character, word, heading and so on). Then swiping down and up moves by that level, and this also draws from the accessible name. So if I want to review someone's name or a URL by character, every period will instead show up as "space d o t space" which takes 5 swipes instead of 1.

Personally, I think the dot substitution causes more problems than it solves, but maybe others prefer it to remain for speech clarity (CC @flameborn @ghorthalon)

**Comment by Flameborn** [comment]
Generally, anything followed by a dot should not make the speech synth pause, unless it is an uppercase letter. So in the example, _mastodon.social_ should not pause speech, unless it is written as _mastodon.Social_.

I think it is perfectly fine if the screen name is mentioned last, it is mostly the content we are interested in, especially when browsing quickly. I don't consider it rude, since there will be ways to see who posted the particular toot, so we certainly do not ignore it.

Generally, when there are multiple columns presented, it is good to have a way to decide what goes first, similarly to how you would click on headers for a table to sort the columns, since this can speed up browsing content significantly. Twitterrific and Dystopia for Reddit are apps that do this, along the official Mastodon web interface and Pinafore, as mentioned before.

These changes would greatly improve the user experience for VO users.

**Comment by DagAgren** [comment]
I'm trying to figure out a good way to present these things, but I keep running into problems here. For instance, I figured maybe it makes more sense to remove screen names entirely from the reading of the toot, since a) it's weird to have it tacked on to the end and b) it's even weirder when the toot is boosted, and there are two names. Instead, the actions could read out both the full name and screen name.

But this has the issue that the toot itself only contains screen names when there are mentions. It seems a bit of a mess to mix them up like that.

**Comment by Flameborn** [comment]
I am still on the opinion that screen names at the end are definitely not weird.

When reading a timeline, it makes sense to announce the full name and the toot text immediately after it, because you are reading linearly, and those are the most important bits that you are interested in first.

Imagine you have an online chat conversation, and you are required to enter your address, along with your full name. You would not want to display those two pieces of information right at the beginning of each paragraph of the chat history, because what is relevant when chatting is the full name (i.e. who sent the message) and what the message is. Reading through someone's address every single time makes it really tedious, because there is no way linearly to move to the next column, skipping whatever is not relevant at that time.

The same goes for reading the received date of an incoming chat message. It might be important on the next day, but when you are chatting, it is unnecessary information, since you are there when the chat message comes in.

In this context, while the screen name on Mastodon is essential to have, its part of someone's identity after all, their full name and the post content is much more important.

If, for some reason, you are still not convinced about this, I think the best solution that can satisfy everybody is to have a way to reorder the information presented when you flick through items. This way, you can set the screen name first, if that is what you are interested in, or have an exotic setup where the post content is always the very last item to be announced.

**Comment by DagAgren** [comment]
The problems with this, though, are:

1. How do you handle boosted toots? Do you put two screen names at the end? That is very confusing.
2. The actual toot texts only contain screen names, not full names, which makes seem to me would make things more confusing.

**Comment by Flameborn** [comment]
The Mastodon web interface, as well as Pinafore handles this as a regular toot, except it puts the screen name of the person who boosted it at the end. Something like:

_full name_, _post text_, _screen name of boosted post's user_, boosted by _screen name of user who boosted the post_, _relative date (i.e. 1 hour ago)_

Ah, I think I see where you are coming from. I think in the case of boosts or mentions, the screen name is more relevant, since it specifically mentions/addresses the user. Even visually, you can have multiple screen names in the toot text. Unless you are not familiar with microblogging, this should not be confusing to users. Even if it is confusing at first, it's just a matter of being aware that while you have a full name you'd like to present yourself as, since this is not unique, the only way to refer to a user is via their screen name.

**Comment by DagAgren** [comment]
Yeah, but visually, you always either show both names, or only the screen name. In this case, we can show only the full name or only the screen name, which breaks the connection between the two.

For now, I think I will try putting it as "full name: toot text, 5 hours ago by @screen name" and "full name, boosted by full name: toot text, 5 hours ago by @screen name, boosted by @screen name". It's a bit clunky still but maybe will work.

(Hmm, I could even sneak in a feature there that the visual UI does not have: It could be ""full name, boosted by full name: toot text, 5 hours ago by @screen name, boosted 3 hours ago by @screen name". Or is that a bit much?)

The actions will also be either "Open @screename" for mentions and "Open full name (@screen name)" for the poster and booster.

How does that sound?

**Comment by DagAgren** [comment]
Removing the screen name from the start of the toot does also make it easier to remove the dot-substitution, because that was added specifically because the reader would pause on some screen names and that was incredibly annoying when it was at the start of a toot, but if it's at the end, I suppose it is no longer as pressing a need.

**Comment by Flameborn** [comment]
You are right, but to be honest, is someone's full name relevant when you are addressing them? I think it is more important to get your message across, at least with mentions, since you are aware of who you are mentioning as a reply to their post, or when sending them a mention outside a thread. In both cases, it is a good idea to know someone's screen name, since you can't address someone by their full name as part of the protocol.

With boosts, I would also go for the screen name, in case the content is coming from a user I am not familiar with and I'd like to follow, mention, etc. So according to your example, I would just rather have: _full name_: toot, _1 hour ago_, boosted by _screen name_. Or, if you think the full name is important, I would put it before the boosted by screen name
, e.g.: boosted by _full name (screen name)_.

I like your idea of knowing the relative date of the boost. It might be a bit too much, but since it's at the end, you can skip past it by swiping to the next item in the timeline.

@pitermach, what do you think?

**Comment by pitermach** [comment]
Sounds good to me. Having the screen names in the actions is also OK, because it matches up with what's being displayed visually with the ScreenName being turned into a highlited link you can click/tap on in clients.

**Comment by DagAgren** [comment]
All right, I have something now, and let's go with that and report back if it is still too annoying or needs tweaking!

(Best to open a new issue in that case)
